### PR TITLE
ddl: unsupport column type change to/from GBK

### LINKED ISSUE #31011: We shouldn’t modify a column‘s charset by `alter table … modify column…`
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

Currently we do not support "alter table…modify column…" to change the character set of a column (except utf8-> UTF8MB4).
```
create table t(a char(25) charset utf8mb4, b char(20)  charset gbk) charset gbk;
alter table t modify column a varchar(255) charset gbk;
create table t1(a varchar(25) charset utf8mb4, b char(20)  charset gbk) charset gbk;
alter table t1 modify column a varchar(255) charset gbk;
```

### 2. What did you expect to see? (Required)
```
tidb>  alter table t modify column a varchar(255) charset gbk;
ERROR 8200 (HY000): Unsupported modify charset from utf8mb4 to gbk
tidb>  alter table t1 modify column a varchar(255) charset gbk;
ERROR 8200 (HY000): Unsupported modify charset from utf8mb4 to gbk
```

### 3. What did you see instead (Required)
```
tidb>  alter table t modify column a varchar(255) charset gbk;
Query OK, 0 rows affected (2.52 sec)
tidb>  alter table t1 modify column a varchar(255) charset gbk;
ERROR 8200 (HY000): Unsupported modify charset from utf8mb4 to gbk
```

### 4. What is your TiDB version? (Required)

```
| Release Version: v5.4.0-alpha-501-g93d28b9d3-dirty
Edition: Community
Git Commit Hash: 93d28b9d3ffe2726640a72342a3c517f88bbd37d
Git Branch: HEAD
UTC Build Time: 2021-12-24 09:13:53
GoVersion: go1.16.4
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false |
```

### PR DESCRIPTION
**Pull request by Defined2014**
### What problem does this PR solve?

Issue Number: close #31011

Problem Summary: not support column type change with GBK charset

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
Not support column type change with GBK charset
```

### CODE REVIEW

File: ddl/ddl_api.go [hunk 1]
```diff
@@ -3926,6 +3926,10 @@ func checkModifyTypes(ctx sessionctx.Context, origin *types.FieldType, to *types
 	}
 
 	err = checkModifyCharsetAndCollation(to.Charset, to.Collate, origin.Charset, origin.Collate, needRewriteCollationData)
+	// TODO, after fully test, we could delete it
```
Comment (zimulala) [inline ddl/ddl_api.go:3929]: Maybe we can remove this comment.

File: ddl/column_test.go [hunk 2]
```diff
@@ -1191,6 +1194,13 @@ func (s *testColumnSuite) TestModifyColumn(c *C) {
 		{"decimal(2,1)", "int", nil},
 		{"decimal", "int", nil},
 		{"decimal(2,1)", "bigint", nil},
+		{"int", "varchar(10) character set gbk", errUnsupportedModifyCharset.GenWithStackByArgs("charset from binary to gbk")},
+		{"varchar(10) character set gbk", "int", errUnsupportedModifyCharset.GenWithStackByArgs("charset from gbk to binary")},
+		{"varchar(10) character set gbk", "varchar(10) character set utf8", errUnsupportedModifyCharset.GenWithStackByArgs("charset from gbk to utf8")},
+		{"varchar(10) character set gbk", "char(10) character set utf8", errUnsupportedModifyCharset.GenWithStackByArgs("charset from gbk to utf8")},
+		{"varchar(10) character set utf8", "char(10) character set gbk", errUnsupportedModifyCharset.GenWithStackByArgs("charset from utf8 to gbk")},
+		{"varchar(10) character set utf8", "varchar(10) character set gbk", errUnsupportedModifyCharset.GenWithStackByArgs("charset from utf8 to gbk")},
+		{"varchar(10) character set gbk", "varchar(255) character set gbk", nil},
```
Comment (tangenta) [inline ddl/column_test.go:1203]: What about GBK to GBK?

**Review by tangenta — APPROVED** [review]

File: ddl/ddl_api.go [hunk 3]
```diff
@@ -3926,6 +3926,9 @@ func checkModifyTypes(ctx sessionctx.Context, origin *types.FieldType, to *types
 	}
 
 	err = checkModifyCharsetAndCollation(to.Charset, to.Collate, origin.Charset, origin.Collate, needRewriteCollationData)
+	if err != nil && to.Charset == charset.CharsetGBK || origin.Charset == charset.CharsetGBK {
+		return err
+	}
 	// column type change can handle the charset change between these two types in the process of the reorg.
 	if err != nil && errUnsupportedModifyCharset.Equal(err) && canReorg {
 		return nil
```
Comment (zimulala) [inline ddl/ddl_api.go:3934]: Could we update this code like ```
```suggestion
	if err != nil {
		if to.Charset == charset.CharsetGBK || origin.Charset == charset.CharsetGBK{
			errors.Trace(err)
		}
		// Column type change can handle the charset change between these two types in the process of the reorg.
		if errUnsupportedModifyCharset.Equal(err) && canReorg {
			return nil
		}
	} 
```
```

**Review by zimulala — APPROVED** [review]
LGTM

**Pull request merged**